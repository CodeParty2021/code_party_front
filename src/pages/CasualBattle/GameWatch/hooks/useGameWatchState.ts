import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import {
  ResultType,
  useFetchResult,
} from "hooks/ResultAPIHooks/useFetchResult";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";

export type IResponse = {
  isAnalyzing: boolean;
  analyzingError: boolean;
  result?: ResultType;
  json?: string;
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: result, fetchResult } = useFetchResult();
  const { data: json, fetchJson } = useFetchJson();
  const navigate = useNavigate();

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchResult(room.info?.analyzingResult.resultId);
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  const _exitBtnHandler = () => {
    navigate("/casual-battle/waiting-room");
  };

  return {
    isAnalyzing: !room.info?.analyzingResult?.resultId,
    analyzingError: Boolean(room.info?.analyzingResult?.error),
    result: result,
    json: json,
    exitBtnHandler: _exitBtnHandler,
  };
};
