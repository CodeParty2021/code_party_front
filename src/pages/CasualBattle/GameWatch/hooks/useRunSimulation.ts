import { useEffect } from "react";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useRunCodes } from "hooks/CodeAPIHooks/useRunCodes";

export type IResponse = {};

export const useRunSimulation = (): IResponse => {
  const { room, isHost, updateRoomInfo } = useRoomSync();
  const {
    data: simulationResult,
    error: simulationError,
    runCodes,
  } = useRunCodes();

  //初期処理
  useEffect(() => {
    //シミュレーションの実行(ホストのみ実行)
    if (isHost) {
      const codes: string[] = [];
      room.sortedKeysOfMembers.forEach((key) => {
        const member = room.members[key];
        if (member.codeId) codes.push(member.codeId);
      });
      runCodes(...codes);
    }
  }, []);

  //シミュレーション結果を受信した場合
  useEffect(() => {
    if (simulationResult) {
      updateRoomInfo({
        analyzingResult: {
          resultId: simulationResult.jsonId,
        },
      });
    }
  }, [simulationResult]);

  //シミュレーション中にエラーが発生した場合
  useEffect(() => {
    if (simulationError) {
      updateRoomInfo({
        analyzingResult: {
          error: "シミュレーション実行中にエラーが発生しました。",
        },
      });
    }
  }, [simulationError]);

  return {};
};
