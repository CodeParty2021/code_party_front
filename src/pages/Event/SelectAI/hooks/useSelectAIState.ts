import { useCodeAPI } from "hooks/CodeAPIHooks/useCodeAPI";
import { EVENT1ON1_INIT_CODE } from "pages/Event/assets/InitCodes";
import { useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { isUser } from "services/user/user";
import { RootState } from "store";

export type IResponse = {
  loading: boolean;
  error: string | undefined;
  beginTrainHandler: (stepId: string) => void;
  backButtonHandler: () => void;
};

export const useSelectAIState = (): IResponse => {
  const { user } = useSelector((state: RootState) => state.user);
  const navigate = useNavigate();
  const { loading, error, createCode, getCodesFilterStepIdAndUserId } =
    useCodeAPI();

  //該当ステップにユーザのコードが存在していればそれをロードするなければ新しく作ってそれをロードする
  const _beginTrainHandler = async (stepId: string) => {
    let codeId: string;
    if (isUser(user)) {
      const codes = await getCodesFilterStepIdAndUserId(stepId, user.id);
      if (codes.length >= 1) {
        codeId = codes[0].id;
      } else {
        const code = await createCode(EVENT1ON1_INIT_CODE, stepId, "1");
        codeId = code.id;
      }
      navigate(`/free-coding/${codeId}/`);
    }
  };

  const _backButtonHandler = () => {
    navigate(`/event/select-mode`);
  };

  return {
    loading: loading,
    error: error,
    beginTrainHandler: _beginTrainHandler,
    backButtonHandler: _backButtonHandler,
  };
};
