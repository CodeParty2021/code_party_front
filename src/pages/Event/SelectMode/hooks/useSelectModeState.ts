import { useCodeAPI } from "hooks/CodeAPIHooks/useCodeAPI";
import { EVENT1ON1_INIT_CODE } from "pages/Event/assets/InitCodes";
import { useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { isUser } from "services/user/user";
import { RootState } from "store";

const EVENT1ON1_STEP_ID = "2";
export type IResponse = {
  loading: boolean;
  error: string | undefined;
  beginTrainHandler: () => void;
};

export const useSelectModeState = (): IResponse => {
  const { user } = useSelector((state: RootState) => state.user);
  const navigate = useNavigate();

  const { loading, error, createCode, getCodesFilterStepIdAndUserId } =
    useCodeAPI();
  //該当ステップにユーザのコードが存在していればそれをロードするなければ新しく作ってそれをロードする
  const _beginTrainHandler = async () => {
    console.log(_beginTrainHandler);
    let codeId: string;
    console.log("isUser(user)", isUser(user));
    console.log(user);
    if (isUser(user)) {
      const codes = await getCodesFilterStepIdAndUserId(
        EVENT1ON1_STEP_ID,
        user.id
      ); //TODO エラー処理
      console.log("getCodesFilterStepIdAndUserId", codes);
      if (codes.length >= 1) {
        codeId = codes[0].id;
      } else {
        const code = await createCode(
          EVENT1ON1_INIT_CODE,
          EVENT1ON1_STEP_ID,
          "1"
        );
        console.log("createCode:" + JSON.stringify(code));
        codeId = code.id;
      }

      navigate(`/free-coding/${codeId}/`);
    }
  };
  return {
    loading: loading,
    error: error,
    beginTrainHandler: _beginTrainHandler,
  };
};
