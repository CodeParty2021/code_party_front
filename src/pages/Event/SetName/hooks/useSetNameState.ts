import { useUserAPI } from "hooks/UserAPIHooks/userAPIHooks";
import { RefObject, useEffect, useRef, useState } from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { updateUserDisplayName } from "services/user/user";

export type IResponse = {
  loading: boolean;
  error: string | undefined;
  nameInputRef: RefObject<HTMLInputElement>;
  startBtnHandler: () => void;
  nameInputHandler: () => void;
  btnDisabled: boolean;
};

export const useSetNameState = (): IResponse => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const nameInputRef = useRef<HTMLInputElement>(null);
  const { error, loading, updateDisplayName } = useUserAPI();
  const [errorDisplay, setErrorDisplay] = useState<string | undefined>();
  const [btnDisabled, setBtnDisabled] = useState<boolean>(true);
  useEffect(() => {
    setErrorDisplay(error);
  }, [error]);
  const _changeNameHandler = () => {
    if (nameInputRef.current?.value) {
      updateDisplayName(nameInputRef.current.value).then((response) => {
        dispatch(updateUserDisplayName(response.displayName));
        navigate("/event/select-mode");
      });
    } else {
      setErrorDisplay("値を入力してください");
    }
  };
  const nameInputHandler = () => {
    if (nameInputRef.current?.value === "") {
      setBtnDisabled(true);
    } else {
      setBtnDisabled(false);
    }
  };
  return {
    loading,
    error: errorDisplay,
    nameInputRef,
    startBtnHandler: _changeNameHandler,
    nameInputHandler,
    btnDisabled,
  };
};
