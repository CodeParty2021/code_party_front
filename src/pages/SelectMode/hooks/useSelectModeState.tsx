import { useNavigate } from "react-router-dom";

export type IResponse = {
  beginDevelopHandler: () => void;
  beginRoomMatchHandler: () => void;
};

export const useSelectModeState = (): IResponse => {
  const navigate = useNavigate();

  //該当ステップにユーザのコードが存在していればそれをロードするなければ新しく作ってそれをロードする
  const _beginDevelopHandler = () => {
    navigate(`/robot-development/top`);
  };

  const _beginRoomMatchHandler = () => {
    navigate("/room-match");
  };

  return {
    beginDevelopHandler: _beginDevelopHandler,
    beginRoomMatchHandler: _beginRoomMatchHandler,
  };
};
