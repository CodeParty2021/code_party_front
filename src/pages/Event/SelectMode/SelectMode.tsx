import React from "react";
import { Background } from "../components/Background";
import { ModeSelectCard } from "./components/ModeSelectCard/ModeSelectCard";
import { useSelectModeState } from "./hooks/useSelectModeState";
import { InnerBox, MarginBox } from "./SelectModeStyle";

type Props = {};

export const EventSelectMode: React.FC<Props> = () => {
  const { loading, beginTrainHandler, beginBattleHandler } =
    useSelectModeState();
  if (loading) {
    return <div>ロード中</div>;
  }
  return (
    <>
      <InnerBox>
        <MarginBox>
          <ModeSelectCard
            mode={"solo"}
            onClick={beginTrainHandler}
          ></ModeSelectCard>
        </MarginBox>
        <MarginBox>
          <ModeSelectCard
            mode={"battle"}
            onClick={beginBattleHandler}
          ></ModeSelectCard>
        </MarginBox>
      </InnerBox>{" "}
      <Background />
    </>
  );
};
