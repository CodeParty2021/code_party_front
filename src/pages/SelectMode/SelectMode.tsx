import React from "react";
import {
  DescriptionText,
  ModeSelectCard,
} from "components/ModeSelectCard/ModeSelectCard";
import { useSelectModeState } from "./hooks/useSelectModeState";
import { InnerBox, MarginBox } from "./SelectModeStyle";

import { StarBackground } from "components/StarBackground/StarBackground";
type Props = {};

export const SelectMode: React.FC<Props> = () => {
  const { beginDevelopHandler, beginRoomMatchHandler } = useSelectModeState();

  return (
    <StarBackground>
      <InnerBox>
        <MarginBox>
          <ModeSelectCard
            imageName="walking_lobo"
            title="ロボット開発モード"
            description={
              <DescriptionText>
                <>
                  開発者として新しくロボットを動かすプログラムを
                  <br />
                  作成する事ができます。
                </>
              </DescriptionText>
            }
            icon="setting"
            onClick={beginDevelopHandler}
            disabled={true}
          ></ModeSelectCard>
        </MarginBox>
        <MarginBox>
          <ModeSelectCard
            imageName="grab_flag_lobo"
            title="フレンド対戦モード"
            description={
              <DescriptionText>
                過去に作成したプログラムで
                <br />
                オンライン対戦することが出来ます。
              </DescriptionText>
            }
            icon="sword"
            onClick={beginRoomMatchHandler}
          ></ModeSelectCard>
        </MarginBox>
      </InnerBox>{" "}
    </StarBackground>
  );
};
