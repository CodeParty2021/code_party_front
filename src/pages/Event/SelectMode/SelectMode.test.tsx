import React from "react";
import { shallow } from "enzyme";
import { EventSelectMode } from "./SelectMode";
import { IResponse, useSelectModeState } from "./hooks/useSelectModeState";

jest.mock("./hooks/useSelectModeState");

const useSelectModeStateMock = useSelectModeState as jest.Mock;

const initialState: IResponse = {
  loading: false,
  error: undefined,
  beginTrainHandler: () => {},
  beginBattleHandler: () => {},
};

describe("<EventSelectMode />", () => {
  beforeEach(() => {
    useSelectModeStateMock.mockReturnValue({ ...initialState });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("auth snapshot test", () => {
    const wrapper = shallow(<EventSelectMode />);

    expect(wrapper.getElements()).toMatchSnapshot();
  });
});
