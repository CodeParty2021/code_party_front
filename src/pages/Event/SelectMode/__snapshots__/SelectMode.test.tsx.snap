// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventSelectMode /> auth snapshot test 1`] = `
Array [
  <React.Fragment>
    <styled.div>
      <styled.div>
        <ModeSelectCard
          mode="solo"
          onClick={[Function]}
        />
      </styled.div>
      <styled.div>
        <ModeSelectCard
          mode="battle"
          onClick={[Function]}
        />
      </styled.div>
    </styled.div>
     
    <Styled(StarBackground) />
  </React.Fragment>,
]
`;
